pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,493 @@
+		if errors.IsNotFound(err) {
+			delete(c.consistentHashMap, eipName)
+			return nil
+		}
+		return err
+	}
+
+	if err := c.syncConsistentHashByEIP(eip); err != nil {
+		klog.ErrorS(err, "Sync consistentHash by EIP failed", "EIP", eip.Name)
+		return err
+	}
+
+	return nil
+}
+
+func newNodeConsistentHashMap() *consistenthash.Map {
+	return consistenthash.New(defaultVirtualNodeReplicas, defaultHashFn)
+}
+
+// syncConsistentHashByEIP refresh the consistentHashMap
+func (c *Cluster) syncConsistentHashByEIP(eip *v1alpha2.ExternalIPPool) error {
+	nodeSel, err := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+	if err != nil {
+		return err
+	}
+	nodes, err := c.nodeLister.List(nodeSel)
+	if err != nil {
+		return err
+	}
+	c.consistentHashMap[eip.Name] = newNodeConsistentHashMap()
+	aliveNodes := c.aliveNodeList()
+	// node alive and node labels matches with externalIPPool nodeSelector
+	aliveAndMatchedNodes := sets.NewString()
+	for _, node := range nodes {
+		nodeName := node.Name
+		if aliveNodes.Has(nodeName) {
+			aliveAndMatchedNodes.Insert(nodeName)
+		}
+	}
+	c.consistentHashMap[eip.Name].Add(aliveAndMatchedNodes.List()...)
+	c.notify(eip.Name)
+	return nil
+}
+
+func (c *Cluster) handleClusterNodeEvents(nodeEvent *memberlist.NodeEvent) {
+	eips, err := c.externalIPPoolLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Handle clusterNode events failed")
+	}
+	switch node, event := nodeEvent.Node, nodeEvent.Event; event {
+	case memberlist.NodeJoin:
+		// when node join cluster, all matched externalIPPools consistentHash should changed
+		node, err := c.nodeLister.Get(node.Name)
+		if err != nil {
+			klog.ErrorS(err, "Processing clusterNode events failed")
+			return
+		}
+		count := 0
+		for _, eip := range eips {
+			if c.matchNodeSelector(node, eip) {
+				c.queue.Add(eip.Name)
+				klog.V(5).InfoS("Node join event", "nodeName", node.String(), "eip", eip.Name)
+				count++
+			}
+		}
+		klog.InfoS("Node event: join node", "nodeName", node.Name, "eipNum", count)
+	case memberlist.NodeLeave:
+		// when node leave cluster, the node may failed or deleted, all externalIPPools consistentHash maybe changed
+		for _, eip := range eips {
+			c.queue.Add(eip.Name)
+			klog.V(5).InfoS("Node leave event", "nodeName", node.String(), "eip", eip.Name)
+		}
+		klog.InfoS("Node event: leave node", "nodeName", node.Name, "eipNum", len(eips))
+	case memberlist.NodeUpdate:
+		klog.InfoS("Node event: update node", "nodeName", node.Name)
+	default:
+		klog.InfoS("Node event: unknown", "nodeName", node.Name, "event", event)
+	}
+}
+
+// aliveNodeList return alive nodeName list in cluster
+func (c *Cluster) aliveNodeList() sets.String {
+	aliveMembers := c.mList.Members()
+	nodes := make([]string, len(aliveMembers))
+	for i, node := range aliveMembers {
+		nodes[i] = node.Name
+	}
+	return sets.NewString(nodes...)
+}
+
+func (c *Cluster) shouldSelect(consistentHash *consistenthash.Map, name string) bool {
+	myNode := c.nodeName
```

Comment:
I feel such local variable is pointless. A point of member variable is that it can be used directly in the member methods. I don't think assigning it to a local variable is helpful.

Reply:
done